pkg/failpoint/fail.go

```diff
@@ -0,0 +1,293 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package failpoint
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Type is the type of failpoint to specifies which action to take.
+type Type int
+
+const (
+	// TypeInvalid is invalid type
+	TypeInvalid Type = iota
+	// TypeOff takes no action
+	TypeOff
+	// TypeError triggers failpoint error with specified argument
+	TypeError
+	// TypePanic triggers panic with specified argument
+	TypePanic
+	// TypeDelay sleeps with the specified number of milliseconds
+	TypeDelay
+)
+
+// String returns the name of type.
+func (t Type) String() string {
+	switch t {
+	case TypeOff:
+		return "off"
+	case TypeError:
+		return "error"
+	case TypePanic:
+		return "panic"
+	case TypeDelay:
+		return "delay"
+	default:
+		return "invalid"
+	}
+}
+
+// Failpoint is used to add code points where error or panic may be injected by
+// user. The user controlled variable will be parsed for how the error injected
+// code should fire. There is the way to set the rule for failpoint.
+//
+// 	<count>*<type>[(arg)][-><more terms>]
+//
+// The <type> argument specifies which action to take; it can be one of:
+//
+//	off:	Takes no action (does not trigger failpoint and no argument)
+//	error:	Triggers failpoint error with specified argument(string)
+//	panic:	Triggers panic with specified argument(string)
+//	delay:	Sleep the specified number of milliseconds
+//
+// The <count>* modifiers prior to <type> control when <type> is executed. For
+// example, "5*error(oops)" means "return error oops 5 times total". The
+// operator -> can be used to express cascading terms. If you specify
+// <term1>-><term2>, it means that if <term1> does not execute, <term2> will
+// be evaluated. If you want the error injected code should fire in second
+// call, you can specify "1*off->1*error(oops)".
+//
+// Based on fail(9) freebsd: https://www.freebsd.org/cgi/man.cgi?query=fail&sektion=9&apropos=0&manpath=FreeBSD%2B10.0-RELEASE
+type Failpoint struct {
```

Comment:
Thanks! Yes, agree to use it as repo-level package.
Since this commit just introduces very simple function, I think we should split it out if the package grows. WDYT? :)

Reply:
SGTM

Reply:
SGTM should also include the original "permissive" license or you could probably stick with the original if moving.

Reply:
Added link about failpoint~